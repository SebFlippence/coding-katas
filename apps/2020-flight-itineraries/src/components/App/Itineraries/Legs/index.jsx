import React from 'react';
import {
  BpkGridContainer,
  BpkGridRow,
  BpkGridColumn,
} from 'bpk-component-grid';

import STYLES from './index.scss';

const formatTime = date => {
  const d = new Date(date);
  const dtf = new Intl.DateTimeFormat('en', {
    year: 'numeric',
    month: 'short',
    day: '2-digit',
    hour12: false,
    hour: 'numeric',
    minute: 'numeric',
  });
  const [, , , , , , { value: hour }, , { value: minute }] = dtf.formatToParts(
    d,
  );

  return `${hour}:${minute}`;
};

const formatMinsToHours = m =>
  `${(m / 60) ^ 0}`.slice(-2) + 'hÂ ' + ('0' + (m % 60)).slice(-2);

const Leg = ({ item }) => (
  <BpkGridContainer fullWidth>
    <BpkGridRow>
      <BpkGridColumn width={1}>
        <img
          src={`/img/airlines/${item.airline_id}.png`}
          width="25"
          title={item.airline_name}
          className={STYLES.airlineImage}
        />
      </BpkGridColumn>
      <BpkGridColumn width={2}>
        {formatTime(item.departure_time)}
        <br />
        <span className={STYLES.airport}>{item.departure_airport}</span>
      </BpkGridColumn>
      <BpkGridColumn width={1.5}>
        <img src="/img/arrow.png"></img>
      </BpkGridColumn>
      <BpkGridColumn width={3}>
        {formatTime(item.arrival_time)}
        <br />
        <span className={STYLES.airport}>{item.arrival_airport}</span>
      </BpkGridColumn>
      <BpkGridColumn width={4} className={STYLES.duration}>
        <small>
          {formatMinsToHours(item.duration_mins)}
          <br />
          {item.stops ? (
            <span className={STYLES.notDirect}>{item.stops} stop</span>
          ) : (
            <span className={STYLES.direct}>Direct</span>
          )}
        </small>
      </BpkGridColumn>
    </BpkGridRow>
  </BpkGridContainer>
);

const Legs = ({ data }) => data.map(leg => <Leg key={leg.id} item={leg} />);

export default Legs;
