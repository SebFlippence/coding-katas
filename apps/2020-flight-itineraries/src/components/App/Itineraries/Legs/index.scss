.airlineImage {
  margin-left:-5px;
}

.airport {
  color: grey;
}

.duration {
  color: grey;
  padding-left: 40px;
}

.direct {
  color: green;
}

.notDirect {
  color: red;
}
