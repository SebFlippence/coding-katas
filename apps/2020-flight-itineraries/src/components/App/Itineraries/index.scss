.card {
  width: 365px;
  margin-left: 10px;
}

.item {
  margin-bottom: 20px;
}

.price {
  font-size: 1.5rem;
}

.agent {
  color: #ccc;
}
