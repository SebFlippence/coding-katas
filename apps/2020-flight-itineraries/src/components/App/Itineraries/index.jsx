import React from 'react';
import BpkCard from 'bpk-component-card';
import BpkButton from 'bpk-component-button';
import {
  BpkGridContainer,
  BpkGridRow,
  BpkGridColumn,
} from 'bpk-component-grid';

import STYLES from './index.scss';
import Legs from './Legs';

const Itinerary = ({ item }) => (
  <BpkGridContainer fullWidth>
    <BpkGridRow>
      <BpkGridColumn width={8}>
        <span className={STYLES.price}>{item.price}</span>
        <br />
        <span className={STYLES.agent}>{item.agent}</span>
      </BpkGridColumn>
      <BpkGridColumn width={1}>
        <BpkButton>Select</BpkButton>
      </BpkGridColumn>
    </BpkGridRow>
  </BpkGridContainer>
);

const Itineraries = ({ flights }) =>
  flights.itineraries.map(itinerary => (
    <section key={itinerary.id} className={STYLES.item}>
      <BpkCard className={STYLES.card}>
        <Legs
          className="Legs"
          data={flights.legs.filter(leg => itinerary.legs.includes(leg.id))}
        />
        <Itinerary item={itinerary} />
      </BpkCard>
    </section>
  ));

export default Itineraries;
