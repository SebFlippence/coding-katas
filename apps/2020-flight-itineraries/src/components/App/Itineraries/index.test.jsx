import React from 'react';
import { render } from '@testing-library/react';

import Itineraries from '.';

describe('Itineraries', () => {
  const mockItineraries = {
    itineraries: [
      {
        id: 'it_1',
        legs: ['leg_1', 'leg_4'],
        price: 'Â£35',
        agent: 'Wizzair.com',
        agent_rating: 9.1,
      },
    ],
    legs: [
      {
        id: 'leg_1',
        departure_airport: 'BUD',
        arrival_airport: 'LTN',
        departure_time: '2020-10-31T15:35',
        arrival_time: '2020-10-31T17:00',
        stops: 0,
        airline_name: 'Wizz Air',
        airline_id: 'WZ',
        duration_mins: 145,
      },
      {
        id: 'leg_4',
        departure_airport: 'LTN',
        arrival_airport: 'BUD',
        departure_time: '2020-11-11T19:45',
        arrival_time: '2020-11-11T21:10',
        stops: 0,
        airline_name: 'Wizz Air',
        airline_id: 'WZ',
        duration_mins: 145,
      },
    ],
  };

  it('renders itineraries correctly', () => {
    const { container } = render(<Itineraries flights={mockItineraries} />);
    expect(container).toMatchSnapshot();
  });
});
