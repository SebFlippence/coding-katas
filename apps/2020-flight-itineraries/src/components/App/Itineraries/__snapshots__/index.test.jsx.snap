// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Itineraries renders itineraries correctly 1`] = `
<div>
  <section
    class="item"
  >
    <div
      class="bpk-card bpk-card--padded card"
      role="button"
    >
      <div
        class="bpk-grid__container bpk-grid__container--full-width"
      >
        <div
          class="bpk-grid__row bpk-grid__row--padded"
        >
          <div
            class="bpk-grid__column bpk-grid__column--1 bpk-grid__column--padded"
          >
            <img
              class="airlineImage"
              src="/img/airlines/WZ.png"
              title="Wizz Air"
              width="25"
            />
          </div>
          <div
            class="bpk-grid__column bpk-grid__column--2 bpk-grid__column--padded"
          >
            15:35
            <br />
            <span
              class="airport"
            >
              BUD
            </span>
          </div>
          <div
            class="bpk-grid__column bpk-grid__column--1.5 bpk-grid__column--padded"
          >
            <img
              src="/img/arrow.png"
            />
          </div>
          <div
            class="bpk-grid__column bpk-grid__column--3 bpk-grid__column--padded"
          >
            17:00
            <br />
            <span
              class="airport"
            >
              LTN
            </span>
          </div>
          <div
            class="bpk-grid__column bpk-grid__column--4 bpk-grid__column--padded duration"
          >
            <small>
              2h 25
              <br />
              <span
                class="direct"
              >
                Direct
              </span>
            </small>
          </div>
        </div>
      </div>
      <div
        class="bpk-grid__container bpk-grid__container--full-width"
      >
        <div
          class="bpk-grid__row bpk-grid__row--padded"
        >
          <div
            class="bpk-grid__column bpk-grid__column--1 bpk-grid__column--padded"
          >
            <img
              class="airlineImage"
              src="/img/airlines/WZ.png"
              title="Wizz Air"
              width="25"
            />
          </div>
          <div
            class="bpk-grid__column bpk-grid__column--2 bpk-grid__column--padded"
          >
            19:45
            <br />
            <span
              class="airport"
            >
              LTN
            </span>
          </div>
          <div
            class="bpk-grid__column bpk-grid__column--1.5 bpk-grid__column--padded"
          >
            <img
              src="/img/arrow.png"
            />
          </div>
          <div
            class="bpk-grid__column bpk-grid__column--3 bpk-grid__column--padded"
          >
            21:10
            <br />
            <span
              class="airport"
            >
              BUD
            </span>
          </div>
          <div
            class="bpk-grid__column bpk-grid__column--4 bpk-grid__column--padded duration"
          >
            <small>
              2h 25
              <br />
              <span
                class="direct"
              >
                Direct
              </span>
            </small>
          </div>
        </div>
      </div>
      <div
        class="bpk-grid__container bpk-grid__container--full-width"
      >
        <div
          class="bpk-grid__row bpk-grid__row--padded"
        >
          <div
            class="bpk-grid__column bpk-grid__column--8 bpk-grid__column--padded"
          >
            <span
              class="price"
            >
              £35
            </span>
            <br />
            <span
              class="agent"
            >
              Wizzair.com
            </span>
          </div>
          <div
            class="bpk-grid__column bpk-grid__column--1 bpk-grid__column--padded"
          >
            <button
              class="bpk-button"
              type="button"
            >
              Select
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
